### PR TITLE
allow passing specific method results

### PR DESCRIPTION
**Pull request by ipa-mdl**
This PR patches the method call system to handle `ua.StatusCode` and `ua.CallMethodResult` properly.
So the callback can indicate errors or invalid arguments.

### CODE REVIEW

**Comment by oroulet** [comment]
so you mean that ua methods should be allowed to return an StatusCode instead of a CallMethodResult? Is this in spec? a CallMethodResult allready contains a StatusCode...

File: opcua/common/methods.py [hunk 1]
```diff
@@ -60,7 +60,11 @@ def wrapper(parent, *args):
             parent = args[0]
             args = args[1:]
             result = func(self, parent, *[arg.Value for arg in args])
-
+        if isinstance(result, ua.CallMethodResult):
+            result.OutputArguments = to_variant(*result.OutputArguments)
+            return result
+        elif isinstance(result, ua.StatusCode):
+            return result
```
[the remark this answers is not in the document]
Comment (ipa-mdl) [reply opcua/common/methods.py]: The client will raise an exception.

**Comment by ipa-mdl** [comment]
> so you mean that ua methods should be allowed to return an StatusCode instead of a CallMethodResult?

This is just a convenience option, it will be put in CallMethodResult.StatusCode automatically.

**Comment by oroulet** [comment]
I am not sure about that one. If you first handle ua objects in your function then I do not see the adavntage of avoiding creating a CallMethodResult. What would be pythonic is to allow to catch exception in decorator, so users can raise exception  instead of returning CallMethodResult with a bad statusCode

**Comment by ipa-mdl** [comment]
I am not sure if we are discussing the same thing..
I'd like to step back and discuss the intended use case. We can figure out the python interface aftewards.

Regarding the specification Part 4 5.11.2:
StatusCode:
 > StatusCode of the Method executed in the server. This StatusCode is set to the Bad_InvalidArgument StatusCode if at least one input argument broke a constraint (e.g. wrong data type, value out of range).
This StatusCode is set to a bad StatusCode if the Method execution failed in the server, e.g. based on an exception.

InputArgumentResults;
>List of StatusCodes corresponding to the inputArguments.
This list is empty unless the operation level result is Bad_InvalidArgument.
If this list is populated, it has the same length as the inputArgument list.

Right now there is no way to set the operation-level StatusCode.
And in addtion `InputArgumentResults` gets filled only if the StatusCode is good.

I'd like to set these from the method callback (somehow), e.g. to notifiy the client that the method failed for a specific reason.
And if the method is decorated, outputArguments should get converted automatically.

I will adapt the PR to remove the special StatusCode handling and just allow passing CallMethodResult directly.

**Comment by oroulet** [comment]
Sorry. Now i understand!  I let you decide what you think is best and i will merge

**Comment by ipa-mdl** [comment]
I think return a StatusCode is the most common use case, so this should be simple.
-> I have removed the other commit again.

**Comment by ipa-mdl** [comment]
I have fixed another issue in handling method results:

In methods without `uamethod` decoration. a list will be kept as-is.
With the decorator the list will get converted into an array variant.
I.e. there was no way to return multiple output arguments with the decorated methods.

This patch treats tuples as multiple arguments, but transform list into an array variant:
```python
return [1,2,3] # OutputArguments:[Variant(val:[1, 2, 3], type:VariantType.Int64)])
return 1,2,3 # OutputArguments:[Variant(val:1, type:VariantType.Int64), Variant(val:2, type:VariantType.Int64), Variant(val:3, type:VariantType.Int64)])

**Pull request merged**